### DISCUSSION TITLE
use different version of Cefsharp when redist application

### DISCUSSION THREAD

**Discussion by mayssaz** [cefsharp/CefSharp#4112] [General]
We use Cefsharp to develop a plugin in order to use our web application in an application. 

When we publish our plugin, We had the following Exception when we have two plugins used a different version of CefSharp

`************** Inner Exception **************
Exception type:System.InvalidOperationException
Exception message:Cef.Initialize() failed.Check the log file see https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#log-file for details.
Exception source:CefSharp.Wpf
Exception Target Site: NoInliningConstructor
   at CefSharp.Wpf.ChromiumWebBrowser.NoInliningConstructor() in C:\projects\cefsharp\CefSharp.Wpf\ChromiumWebBrowser.cs:line 509
   at CefSharp.Wpf.ChromiumWebBrowser..ctor() in C:\projects\cefsharp\CefSharp.Wpf\ChromiumWebBrowser.cs:line 440

************** Outer Exception **************

Exception type:System.Windows.Markup.XamlParseException
Exception message:'The invocation of the constructor on type 'CefSharp.Wpf.ChromiumWebBrowser' that matches the specified binding constraints threw an exception.' Line number '14' and line position '22'.
Exception source:PresentationFramework
Exception Target Site: RewrapException
   at System.Windows.Markup.XamlReader.RewrapException(Exception e, IXamlLineInfo lineInfo, Uri baseUri)
   at System.Windows.Markup.WpfXamlLoader.Load(XamlReader xamlReader, IXamlObjectWriterFactory writerFactory, Boolean skipJournaledProperties, Object rootObject, XamlObjectWriterSettings settings, Uri baseUri)
   at System.Windows.Markup.WpfXamlLoader.LoadBaml(XamlReader xamlReader, Boolean skipJournaledProperties, Object rootObject, XamlAccessLevel accessLevel, Uri baseUri)
   at System.Windows.Markup.XamlReader.LoadBaml(Stream stream, ParserContext parserContext, Object parent, Boolean closeStream)
'

-------
- **What version of the product are you using?**
     -  Release [v101.0.180](https://github.com/cefsharp/CefSharp/releases/tag/v101.0.180)
     - Release [v100.0.230](https://github.com/cefsharp/CefSharp/releases/tag/v100.0.230)
   
- **What architecture x86 or x64?**
x64

- **On what operating system?**
    - Windows Server 2019

- **Are you using `WinForms`, `WPF` or `OffScreen`?** WPF

have you any idea, to specify in which version of dll must be used in each plugin ?

**Comment by amaitland** [comment]
In future please use discussions to ask questions. https://github.com/cefsharp/CefSharp/discussions

It's impossible to load two different versions of CEF in the same process. You would need to host the browser out of process with something like https://github.com/red-gate/RedGate.AppHost to use a different version. 

> When we publish our plugin, We had the following Exception when we have two plugins used a different version of CefSharp

Are you the developer of both plugins? Might be possible if you are to use the already loaded version.

**Reply by mayssaz** [reply]
Hello,

For this test we are the developer of both plugin. 
The final plugin will be published in the marketplace of the host application. So, in the host application, the user can download and install our plugin with a version of Cefsharp and he can use another plugin published from other developers and developed with another version of the library. So we want to avoid this exception. In the installer of our plugin we will add all the necessary libraries for the redistribute. 
I'm newbie in C# (I'm a linux c++ developer). Usually, in linux C++ application, we can force to load a version of library with export LD_LIBRARY_PATH. I don't know if there is an equivalent in C# windows application?

**Reply by amaitland** [reply]
Running the browser out of process is the best option. There's an example at https://github.com/stever/AppHostCefSharp (older by should give you the idea).

**Reply by amaitland** [reply]
There's a work in process for a `CefSharp` specific out of process implementation now available at https://github.com/cefsharp/CefSharp.OutOfProcess
